### ISSUE TITLE
ContinuousQuadratureDataTransfer with non-fixed-size quadrature data structures

### ISSUE THREAD

**Issue by rezarastak** [dealii/dealii#8506]
After today's talk about hp-adaptivity at the workshop, I had a conversion with @peterrum and @marcfehling about using `ContinuousQuadratureDataTransfer` where the data type stored at quadrature points might be different from a cell to another. However, the current implementation of packing and unpacking functions within `ContinuousQuadratureDataTransfer` enforces that all the cells in the triangulation have data structures of the same fixed size (returned by `DataType::number_of_values()`). I assume `ContinuousQuadratureDataTransfer` was written at a time when variable size cell data transfer between subdomains were not fully implemented. However, @marcfehling believes, because of recent advancements in data transfer algorithms made for hp parallel algorithms, we can have variable sized data transfer without any noticeable performance overhead.

What do you think, @jppelteret ? Do you think this is something possible and desirable?

In my structural FE simulations, a lot of times most of the mesh contains elastic materials which don't need any history data accosiated with them. But a small portion of the mesh contains a more complicated material with plasticity/damage/etc that requires complicated storage of many history variables per quadrature point. Currently, using the helpful facilities in `ContinuousQuadratureDataTransfer` and `CellDataStorage`, I have to make sure all cells in the mesh, even the elastic ones, are initialized with history data types that all return the same value from the method `number_of_values()`. The elastic cells, at every refinement step, transfer and interpolate multiple trivial zeros to initialize child members. I have to make these unnecessary data transfers and use strange workaround to allow my code to use multiple materials across an adaptively refined mesh. My objective is to allow any cell in the mesh:

- [ ]  to have no structure attached to it.
- [ ]  to have a structure attached to it which may differ in size between cell to cell.
- [ ]  to have a structure that may grow in time. For example, a cell might be pristine for most of the simulation with elastic response and suddenly cracks after the mechanical stress reaches a critical value. Now the cell needs to store damage and crack opening data at every quadrature point.

**Comment by peterrum** [comment]
Let's setup tomorrow a test!

**Comment by marcfehling** [comment]
We introduced the feature of variable size data transfer in #6864. I did not touch the `ContinuousQuadratureDataTransfer` class at that time because I did not know much about its functionality (and I still don't know more today).

---

Objects of type `FullMatrix` will be transferred from one cell to another.
https://github.com/dealii/dealii/blob/23aeabb0d573fc4cb23c362622be80287cd8c987/include/deal.II/base/quadrature_point_data.h#L854

---

Does the size of these `FullMatrix` objects vary for each cell? I do not see (yet) any hint in the current implementation where this would be the case. Thus we would probably do better with sticking to fixed size transfer.

We first thought the following line of code would motivate the use of variable size transfer, but the documentation suggests that it is meant to supply `number_of_values` to processes that do not have any cell at all.
https://github.com/dealii/dealii/blob/23aeabb0d573fc4cb23c362622be80287cd8c987/include/deal.II/base/quadrature_point_data.h#L800

---

Any thoughts on that? -- @davydden @jppelteret

**Comment by marcfehling** [comment]
Does this class work with p adaptive methods and thus with different quadrature rules on different cells?

**Comment by davydden** [comment]
> Does this class work with p adaptive methods and thus with different quadrature rules on different cells?

At the time it was written hp and p::d::Tria were mutually exclusive, so I guess no ;-)

**Comment by davydden** [comment]
@marcfehling @rezarastak 

> it is meant to supply number_of_values to processes that do not have any cell at all.

all that class does it take provide interface form quadrature point data to feed scalar, in total `number_of_values` on **each** cell, do L2 projection of each scalar on parent, send the effectively DG field to child and evaluate at quadrature point of child. 

> the data type stored at quadrature points might be different from a cell to another

`TransferableQuadraturePointData::number_of_values()` currently does not know anything about each cell, so it does not support varying number of data per cell regardless of the underlying FE space, be it hp or not hp. 

That is, the question of how much is stored per cell (i.e. a scalar or a scalar and a vector) and the question whether or not hp::DoFHandler is involved are separate issue and neither is currently supported.

> Do you think this is something possible and desirable?

certainly possible ;-)

**Comment by jppelteret** [comment]
> Do you think this is something possible and desirable?

@rezarastak Absolutely. Your usage case is one that I'm very familiar with, so I'm very much in favour of the the extensions / refactoring / reimplementation of the `ContinuousQuadratureDataTransfer` class. Whether its possible to adjust the current implementation or if it needs a complete rewrite would require a little bit of investigation, but I wouldn't be surprised if it could be tweaked to support `hp::DoFHandler` and a non-constant data size. I've been following the additions that @marcfehling has been doing with exactly this in mind (but, due to time constraints, its not something that I'll be able to dig into myself). To have a completely flexible way of specifying the amount of data and how its transferred (interpolation, min/max, average etc.) would be really cool, because its exactly these exotic schemes that become really handy when dealing with coarsening/refinement and materials with internal variables.

Ideally I think that this class, or another one like it, could support `p::s::Tria` and `p::d::Tria`, and both normal and `hp::` `DoFHandler`s now that the flexible data transfer mechanisms exist. Even if you were not to provide a full implementation for all possible permutations of `Triangulations` and `DoFHandlers`, it would be good to keep this in mind for `ContinuousQuadratureDataTransfer` version 1.1 or 2.

Maybe @marcfehling could remind me about the status of data transfer and `p::s::Tria` -- did you do anything there?

**Comment by rezarastak** [comment]
I am writing a PR related to this. As part of it I have a to make a decision.

 If a material model does not require quadrature point data, such as an elastic solid, we don't need to even initialize the `CellDataStorage` for the cells with that material id. In that case, we need a way to tell whether a `CellDataStorage` object contains data for a specific `cell` or not. What is the best approach?
1. `CellDataStorage::get_data<T>(cell)` returns empty vector if no data is available for the cell. Current behavior of this method in this situation is assertion failure.
1. We add a public method `CellDataStorage::has_cell(cell)` or `CellDataStorage::contains_cell(cell)`
1. We require that all cells within a triangulation have quadrature point data associated with them. The cells corresponding to elastic materials have to be initialized by an empty struct.

**Comment by marcfehling** [comment]
> Maybe @marcfehling could remind me about the status of data transfer and `p::s::Tria` -- did you do anything there?

@jppelteret  Variable size data transfer was only an issue with `parallel::distributed::Triangulations`. There is no need for transfering variable size data for `parallel::shared::Triangulation` objects since all participating processes have access to all data. `SolutionTranser` and `CellDataTransfer` work with `p::s::Triangulation` as well.

---
> 1. `CellDataStorage::get_data<T>(cell)` returns empty vector if no data is available for the cell. Current behavior of this method in this situation is assertion failure.

@rezarastak Our variable data transfer algorithm can handle cells with no data assigned. Which kind of assertion is triggered. Could you be more specific?

**Comment by rezarastak** [comment]
> Which kind of assertion is triggered? Could you be more specific?

@marcfehling I meant this assertion in `get_data()` method of `CellDataStorage`

https://github.com/dealii/dealii/blob/954642f946d3d456b555ff16f43e2e3adb84be02/include/deal.II/base/quadrature_point_data.h#L667

We have to remove this assertion if we decide to go with option 1.

**Comment by marcfehling** [comment]
> https://github.com/dealii/dealii/blob/954642f946d3d456b555ff16f43e2e3adb84be02/include/deal.II/base/quadrature_point_data.h#L667
> 
> We have to remove this assertion if we decide to go with option 1.

Ah, I see. You may remove the assertion if you rather consider a conditional checking `it != map.end()` and return an empty vector if `it == map.end()`.

Give it a try! Awesome work on this PR :)

**Comment by jppelteret** [comment]
> We add a public method CellDataStorage::has_cell(cell) or CellDataStorage::contains_cell(cell)

In contrast to @marcfehling, I think that having a `stores_data()` method [1] should be the preferred approach. There's a difference between checking to see if the map contains the key (i.e. has been initialised for this cell) and whether or not there is a non-zero amount of data stored on the cell (e.g. a deactivated cell or incorrectly initialised cell). 

[1] We do similar for the [`GeneralDataStorage`](https://dealii.org/developer/doxygen/deal.II/classGeneralDataStorage.html#ab3e62cafaae697b2c6dc6f61d462068f) class.

**Comment by rezarastak** [comment]
@jppelteret your reasoning for using option 2 makes sense and I am working on implementing it. It has one disadvantage though. Running the expression
```c++
if (cell_data_storage.stores_data(cell))
  q_data = cell_data_storage.template get_data<DataType>(cell);
```
will redundantly call `map::find()` function twice (once in `stores_data()` and once in `get_data()`) even in release mode.

**Comment by jppelteret** [comment]
I agree that calling `stores_data()` followed by `get_data()` leads to two lookups in the map. For the purpose of _counting_ the number of stored entries then we could have a more specialised function (admittedly at the expense of a little bit of code duplication):
```c++
std::size_t 
CellDataStorage<>::n_stored_data(const CellIteratorType &cell) const
{
  const auto it = map.find(cell);
  if (it == map.end()) 
    return 0;
  else
    // This could return 0 or something else.
    // At this point we don't care if the data has been incorrectly 
    // initialised or not; that's not what this function is supposed 
    // to check.
    return it->second.size();
}
```
For user code, or other points in our own code, we could bypass the double check in release mode by using the following idiom (where we know for sure that we expect data to be stored):
```c++
Assert (cell_data_storage.stores_data(cell), ExcMessage("Expected data to be stored for this cell."));
q_data = cell_data_storage.template get_data<DataType>(cell);
```

Would this work with what you had in mind?

**Comment by bangerth** [comment]
Alternatively, you could let the `get_data()` function return a `boost::optional` object. That's an incompatible change, or maybe one could just deprecate the old function and instead introduce a new function `get_data_for_cell(const cell_iterator &)` that does so. Returning an empty `boost::optional` object then simply indicates that the cell has no data stored.
